10x/wdls/count.wdl

```diff
@@ -0,0 +1,1072 @@
+    input:
+      sorted_AA=sort_AA.default, sorted_AT=sort_AT.default, sorted_AC=sort_AC.default, sorted_AG=sort_AG.default,
+      sorted_TA=sort_AA.default, sorted_TT=sort_AT.default, sorted_TC=sort_AC.default, sorted_TG=sort_AG.default,
+      sorted_CA=sort_AA.default, sorted_CT=sort_AT.default, sorted_CC=sort_AC.default, sorted_CG=sort_AG.default,
+      sorted_GA=sort_AA.default, sorted_GT=sort_AT.default, sorted_GC=sort_AC.default, sorted_GG=sort_AG.default,
+      sorted_null=sort_null.default,
+      total_AA=sort_AA.total_reads, total_AT=sort_AT.total_reads, total_AC=sort_AC.total_reads, total_AG=sort_AG.total_reads,
+      total_TA=sort_AA.total_reads, total_TT=sort_AT.total_reads, total_TC=sort_AC.total_reads, total_TG=sort_AG.total_reads,
+      total_CA=sort_AA.total_reads, total_CT=sort_AT.total_reads, total_CC=sort_AC.total_reads, total_CG=sort_AG.total_reads,
+      total_GA=sort_AA.total_reads, total_GT=sort_AT.total_reads, total_GC=sort_AC.total_reads, total_GG=sort_AG.total_reads,
+      total_null=sort_null.total_reads
+  }
+  
+
+  call mark_duplicates_split {
+    input: input_bam=sort_by_bc_join.default
+  }
+  
+  scatter(split_file in mark_duplicates_split.split_files) {
+    call mark_duplicates_main {
+      input: input_bam=sort_by_bc_join.default, split_file=split_file, align=extract_reads_join.out_align
+    }
+  }
+  
+  call mark_duplicates_join {
+    input: output_output=mark_duplicates_main.output_bam, chunked_reporter_output=mark_duplicates_main.chunked_reporter
+  }
+
+  call count_genes_split {
+    input: bam_inputs=mark_duplicates_main.output_bam, barcode_whitelist=setup_chunks.barcode_whitelist, gem_groups=extract_reads_join.gem_groups,
+      reference_path=reference_path, barcode_summary=attach_bcs_and_umis_join.barcode_summary, chemistry_def=setup_chunks.chemistry_def
+  }
+
+  scatter(split_file in count_genes_split.split_files) {
+    call count_genes_main {
+    input: bam_inputs=mark_duplicates_main.output_bam, barcode_whitelist=setup_chunks.barcode_whitelist, gem_groups=extract_reads_join.gem_groups,
+      reference_path=reference_path, barcode_summary=attach_bcs_and_umis_join.barcode_summary, chemistry_def=setup_chunks.chemistry_def,
+      split_file=split_file, align=extract_reads_join.out_align
+    }
+  }
+
+  call count_genes_join {
+    input: matrices_h5_output=count_genes_main.matrices_h5, chunked_reporter_output=count_genes_main.chunked_reporter,
+      chemistry_def=setup_chunks.chemistry_def, gem_groups=extract_reads_join.gem_groups, sample_id=sample_id
+  }
+
+  call filter_barcodes_main {
+    input: sample_id=sample_id, matrices_h5=count_genes_join.matrices_h5, barcode_whitelist=setup_chunks.barcode_whitelist,
+      gem_groups=extract_reads_join.gem_groups, chemistry_def=setup_chunks.chemistry_def,
+      raw_fastq_summary=extract_reads_join.summary, attach_bcs_summary=attach_bcs_and_umis_join.summary,
+      barcode_summary=attach_bcs_and_umis_join.barcode_summary
+  }
+}
```

Comment:
Just curious, how long does this take to run on cromwell vs running the native pipeline?

Reply:
It's pretty similar. There are a couple of differences. First, cromwell makes a new copy of the reference_path for every task that has it as an input. So when align_reads starts, you have to wait while it copies ~100GB of identical reference data. And, cromwell seems to want to maximize parallelization, so it will run all instances of STAR at once, even though each instance loads a private copy of the SA into memory.

Reply:
Interesting & good to know. Thanks!